File: Valerio/Python/max_sub_array.py
from random import randint
from time import time

def max_sub_array(a):
    change = [0]
    max = 0
    maxIndex = ()
    sums = []
    for i in range(len(a) - 1):
        if a[i] * a[i + 1] < 0:
            change.append(i + 1)
    change.append(len(a))
    for i in range(len(change) - 1):
        sum = 0
        for j in range(change[i], change[i + 1]):
            sum += a[j]
        sums.append(sum)
    for i in range(len(sums)):
        sum = 0
        for j in range(i, len(sums)):
            sum += sums[j]
            if sum > max:
                max = sum
                maxIndex = (change[i],change[j + 1])

    return a[maxIndex[0]:maxIndex[1]]


a = [randint(-100,100) for i in range(50000)]

t = -time()
max_sub_array(a)
t += time()

print(t)

File: Valerio/Algoritmi/java/src/MaxSubarray.java
import java.util.*;

public class MaxSubarray {
    private final int[] A;

    public MaxSubarray(int[] a) {
        A = a;
    }

    public int[] run() {
        int max = 0, sum;
        int start = 0, end = 0;
        ArrayList<Integer> change = new ArrayList<>();
        ArrayList<Integer> sums = new ArrayList<>();

        change.add(0);

        for (int i = 0; i < A.length - 1; i++)
            if (A[i] * A[i + 1] < 0)
                change.add(i + 1);

        change.add(A.length);

        for (int i = 0; i < change.size() - 1; i++) {
            sum = 0;
            for (int j = change.get(i); j < change.get(i + 1); j++)
                sum += A[j];
            sums.add(sum);
        }

        for (int i = 0; i < sums.size(); i++) {
            sum = 0;
            for (int j = i; j < sums.size(); j++) {
                sum += sums.get(j);
                if (sum > max) {
                    max = sum;
                    start = change.get(i);
                    end = change.get(j + 1);
                }
            }
        }

        return Arrays.copyOfRange(A, start, end);
    }
}